generate source map for uglify if devtool === #source-map

aldarund opened the pull request:
Fix for https://github.com/nuxt/nuxt.js/issues/3450

aldarund committed:
generate source map for uglify if devtool === #source-map

clarkdo commented on lib/builder/webpack/client.js:
```diff
@@ -164,6 +144,27 @@ export default class WebpackClientConfig extends WebpackBaseConfig {
       }
     }
 
+    // Make uglifyjs faster
```
Thank you for you contribution！
Could you explain why move down the plugin ？

aldarund replied:
Because source-map usually ( or at least can be ) added in build extend. E.g. to only build it for client. It is even showed in nuxt documentation https://nuxtjs.org/api/configuration-build/#extend
So i moved it down to be after nuxt evaluate extend function.

clarkdo replied:
Make sense, `!config.optimization.minimizer` check will also prevent overriding user changes in `minimizer`.

And even user deleted `minimizer` in extend, webpack will still give it a default one, we actually do the same thing but with more configs.

clarkdo commented on lib/builder/webpack/client.js:
```diff
@@ -164,6 +144,27 @@ export default class WebpackClientConfig extends WebpackBaseConfig {
+    // Make uglifyjs faster
+    if (!this.options.dev && !config.optimization.minimizer) {
+      // https://github.com/webpack-contrib/uglifyjs-webpack-plugin
+      const sourceMap = config.devtool === '#source-map'
```
Should the value be source-map?
https://webpack.js.org/configuration/devtool/

aldarund replied:
Well, i`m not really sure here.  
https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsApply.js#L236
Seems like it affect only a comment string for a source map and it will work with # @ and without any of it.

aldarund committed:
support different variants of source-map option

clarkdo commented on lib/builder/webpack/client.js:
```diff
@@ -164,6 +144,28 @@ export default class WebpackClientConfig extends WebpackBaseConfig {
+    if (!this.options.dev && !config.optimization.minimizer) {
+      // https://github.com/webpack-contrib/uglifyjs-webpack-plugin
+      const sourceMapVariants = ['#source-map', '@source-map', 'source-map']
+      const sourceMap = sourceMapVariants.includes(config.devtool)
```
The official way should be more accurate.
https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsDefaulter.js#L294

aldarund replied:
@clarkdo hm, i though about implementing same thing as you linked, but it will match for example also #eval-source-map and all other source-map variants. And because when i tested with #eval-source-map - all seems to be working even without this pr ( as described in linked issue), so i didnt go that way. 
but since its in official repo i will change it

clarkdo replied:
Thank you.

aldarund committed:
change source map to be likes in official webpack repo
https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsDefaulter.js#L294

clarkdo commented on lib/builder/webpack/client.js:
```diff
@@ -164,6 +144,27 @@ export default class WebpackClientConfig extends WebpackBaseConfig {
+    // Make uglifyjs faster
+    if (!this.options.dev && !config.optimization.minimizer) {
+      // https://github.com/webpack-contrib/uglifyjs-webpack-plugin
+      const sourceMap = config.devtool && /source-?map/.test(config.devtool)
```
Since `sourceMap` is only used blow, so how about ?
```js
sourceMap: config.devtool && /source-?map/.test(config.devtool)
```

clarkdo reviewed (approved):
LGTM
//cc @Atinux @pi0

aldarund committed:
small code cleanup

aldarund commented:
@Atinux @pi0 ping. can we get this merged please?

aldarund commented:
Whats up with this pr.. Will it be merged  ? :)

Atinux commented:
Thanks @aldarund 

Sorry for the delay
